llvm/lib/SYCL/KernelPropGen/KernelPropGen.cpp

```diff
@@ -0,0 +1,147 @@
+//===- KernelPropGen.cpp                                      ---------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+// Retrieves the names of the kernels inside of the passed in file and places
+// them into a text file. Possible to merge this into another pass if
+// required, as it's a fairly trivial pass on its own.
+// ===---------------------------------------------------------------------===//
+
+#include <cstddef>
+#include <regex>
+#include <string>
+
+#include "llvm/SYCL/KernelPropGen.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Module.h"
+#include "llvm/Pass.h"
+#include "llvm/IR/Function.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+
+// Put the code in an anonymous namespace to avoid polluting the global
+// namespace
+namespace {
+
+/// Retrieve the names for all kernels in the module and place them into a file
+struct KernelPropGen : public ModulePass {
+
+  static char ID; // Pass identification, replacement for typeid
+
+  KernelPropGen() : ModulePass(ID) {}
+
+  bool isKernel(const Function &F) {
+    if (F.getCallingConv() == CallingConv::SPIR_KERNEL)
+      return true;
+    return false;
+  }
+
+  int GetWriteStreamID(StringRef Path) {
+    int FileFD = 0;
+    std::error_code EC =
+          llvm::sys::fs::openFileForWrite(Path, FileFD);
+    if (EC) {
+      llvm::errs() << "Error in KernelPropGen Pass: " << EC.message() << "\n";
+    }
+
+    return FileFD;
+  }
+
+  /// Visit all the functions of the module
+  bool runOnModule(Module &M) override {
+    SmallString<256> TDir;
+    llvm::sys::path::system_temp_directory(true, TDir);
+    std::string file = "KernelProperties_" + M.getSourceFileName();
+    llvm::sys::path::append(TDir, file);
+    llvm::sys::path::replace_extension(TDir, "sh",
```

Comment:
If you are using some Bash extensions, clearer to use `.bash` instead?
But I do not know if you are using some `bash`-ism...